File: scripts/python/parsers/custom_permissions_parser.py
"""Parser for Custom Permission metadata (.customPermission-meta.xml)."""
import re
from pathlib import Path


def _extract_tag(xml: str, tag: str) -> str | None:
    m = re.search(rf'<{tag}[^>]*>([\s\S]*?)</{tag}>', xml)
    return m.group(1).strip() if m else None


def parse_custom_permissions(type_dir: str) -> dict:
    """
    @param type_dir: path to the customPermissions/ folder
    @returns: { permissions: list }
    """
    permissions = []
    p = Path(type_dir)
    if not p.exists():
        return {'permissions': permissions}

    for f in sorted(p.glob('*.customPermission-meta.xml')):
        xml = f.read_text(encoding='utf-8')
        permissions.append({
            'apiName':            f.name.replace('.customPermission-meta.xml', ''),
            'label':              _extract_tag(xml, 'label'),
            'description':        _extract_tag(xml, 'description'),
            'isSessionActivated': _extract_tag(xml, 'isSessionActivated') == 'true',
        })

    return {'permissions': permissions}

File: scripts/node/parsers/customPermissionsParser.js
'use strict';

const fs   = require('fs');
const path = require('path');

function extractTag(xml, tag) {
  const m = xml.match(new RegExp(`<${tag}[^>]*>([\\s\\S]*?)<\\/${tag}>`));
  return m ? m[1].trim() : null;
}

/**
 * Reads Custom Permission metadata from SFDX source format.
 * @param {string} typeDir - path to the customPermissions/ folder
 * @returns {{ permissions: Array }}
 */
async function parseCustomPermissions(typeDir) {
  const permissions = [];
  if (!fs.existsSync(typeDir)) return { permissions };

  const files = fs.readdirSync(typeDir).filter(f => f.endsWith('.customPermission-meta.xml'));

  for (const file of files) {
    const xml = fs.readFileSync(path.join(typeDir, file), 'utf8');
    permissions.push({
      apiName:            file.replace('.customPermission-meta.xml', ''),
      label:              extractTag(xml, 'label'),
      description:        extractTag(xml, 'description'),
      isSessionActivated: extractTag(xml, 'isSessionActivated') === 'true',
    });
  }

  return { permissions };
}

module.exports = { parseCustomPermissions };
